File: java/src/main/java/dev/medzik/libcrypto/Aes.java
package dev.medzik.libcrypto;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.spec.AlgorithmParameterSpec;

public final class Aes {
    private static final String ALGORITHM = "AES";

    public static final AesType GCM = AesType.GCM;
    public static final AesType CBC = AesType.CBC;

    /**
     * Encrypts the given clear bytes using AES with the given key and random IV.
     *
     * @param type AES type to use
     * @param key secret key to use for encryption
     * @param clearBytes clear bytes to encrypt (UTF-8)
     * @return Encrypted cipher (hex encoded)
     * @throws AesException if encryption fails
     */
    public static String encrypt(AesType type, byte[] key, byte[] clearBytes) throws AesException {
        // generate random IV
        byte[] iv = Random.randBytes(type.getIvLength());

        SecretKeySpec secretKey = new SecretKeySpec(key, ALGORITHM);
        AlgorithmParameterSpec parameterSpec = getParameterSpec(type, iv);

        // initialize cipher
        Cipher cipher;
        try {
            cipher = Cipher.getInstance(type.getMode());
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, parameterSpec);
        } catch (Exception e) {
            throw new AesException("Failed to init cipher", e);
        }

        // encrypt
        byte[] cipherBytes;
        try {
            cipherBytes = cipher.doFinal(clearBytes);
        } catch (Exception e) {
            throw new AesException("Failed to finalize encryption", e);
        }

        // return IV + cipher text as hex string
        return Hex.encode(iv) + Hex.encode(cipherBytes);
    }

    /**
     * Decrypts the given cipher text using AES with the given key.
     *
     * @param type AES type to use
     * @param key secret key to use for decryption
     * @param cipherText cipher text to decrypt (hex encoded)
     * @return Decrypted bytes
     * @throws AesException if decryption fails
     */
    public static byte[] decrypt(AesType type, byte[] key, String cipherText) throws AesException {
        // get IV length in hex string
        int ivLength = type.getIvLength() * 2;

        // extract IV and Cipher Text from hex string
        byte[] iv = Hex.decode(cipherText.substring(0, ivLength));
        byte[] cipherBytes = Hex.decode(cipherText.substring(ivLength));

        SecretKeySpec secretKey = new SecretKeySpec(key, ALGORITHM);
        AlgorithmParameterSpec parameterSpec = getParameterSpec(type, iv);

        // initialize cipher
        Cipher cipher;
        try {
            cipher = Cipher.getInstance(type.getMode());
            cipher.init(Cipher.DECRYPT_MODE, secretKey, parameterSpec);
        } catch (Exception e) {
            throw new AesException("Failed to init cipher", e);
        }

        // decrypt
        try {
            return cipher.doFinal(cipherBytes);
        } catch (Exception e) {
            throw new AesException("Failed to finalize decryption", e);
        }
    }

    private static AlgorithmParameterSpec getParameterSpec(AesType type, byte[] iv) {
        switch (type) {
            case CBC:
                return new IvParameterSpec(iv);
            case GCM:
                return new GCMParameterSpec(128, iv);
        }

        // never happen
        return null;
    }

    public enum AesType {
        CBC("AES/CBC/PKCS5Padding", 16),
        GCM("AES/GCM/NoPadding", 12);

        private final String mode;
        private final int ivLength;

        AesType(String mode, int ivLength) {
            this.mode = mode;
            this.ivLength = ivLength;
        }

        private String getMode() {
            return mode;
        }

        private int getIvLength() {
            return ivLength;
        }
    }
}

File: javascript/src/aes.ts
import { Buffer } from "buffer/"
import { createCipheriv, createDecipheriv } from "browserify-cipher"

import * as random from "./random"

const AES_CBC_MODE = 'aes-256-cbc'
const AES_GCM_MODE = 'aes-256-gcm'

/**
 * Encrypts a string using AES CBC mode
 * @param secretKey secret key to use for encryption (hex encoded)
 * @param clearText clear text to use for encryption
 * @return string cipher text (hex encoded)
 */
export function encryptAesCbc(secretKey: string, clearText: string): string {
    return encrypt(secretKey, clearText, AES_CBC_MODE, 16, false)
}

/**
 * Encrypts a string using AES GCM mode
 * @param secretKey secret key to use for encryption (hex encoded)
 * @param clearText clear text to use for encryption
 * @return string cipher text (hex encoded)
 */
export function encryptAesGcm(secretKey: string, clearText: string): string {
    return encrypt(secretKey, clearText, AES_GCM_MODE, 12, true)
}

/**
 * Decrypts the given cipher text using AES CBC mode with the given key.
 * @param secretKey secret key to use for encryption (hex encoded)
 * @param cipherText cipher text to decrypt (hex encoded)
 * @return string clear text (decrypted text)
 */
export function decryptAesCbc(secretKey: string, cipherText: string): string {
    return decrypt(secretKey, cipherText, AES_CBC_MODE, 16, false)
}

/**
 * Decrypts the given cipher text using AES GCM mode with the given key.
 * @param secretKey secret key to use for encryption (hex encoded)
 * @param cipherText cipher text to decrypt (hex encoded)
 * @return string clear text (decrypted text)
 */
export function decryptAesGcm(secretKey: string, cipherText: string): string {
    return decrypt(secretKey, cipherText, AES_GCM_MODE, 12, true)
}

function encrypt(secretKey: string, clearText: string, mode: string, ivLength: number, authMode: boolean) {
    // decode the secret key from a hex string to a buffer
    const key = Buffer.from(secretKey, 'hex')
    // generate a random initialization vector
    let iv = random.randBytes(ivLength)

    // create a cipher using the secret key and the initialization vector
    const cipher = createCipheriv(mode, key, iv)
    // update the cipher with the clear text
    let cipherText = cipher.update(clearText, 'utf8', 'hex')
    // finalize the cipher
    cipherText += cipher.final('hex')

    // add the initialization vector to the cipher text
    cipherText = Buffer.from(iv).toString('hex') + cipherText
    // if authentication is enabled, add auth tag to the cipher
    if (authMode) {
        cipherText += cipher.getAuthTag().toString('hex')
    }

    return cipherText
}

function decrypt(secretKey: string, cipherText: string, mode: string, ivLength: number, authMode: boolean) {
    // decode the secret key from a hex string to a buffer
    const key = Buffer.from(secretKey, 'hex')
    // get the initialization vector from the cipher text
    const iv = Buffer.from(cipherText.substring(0, ivLength * 2), 'hex')
    // remove the initialization vector from the cipher text
    cipherText = cipherText.substring(ivLength * 2)
    let authTag;
    if (authMode) {
        // get the auth tag from the cipher text
        authTag = Buffer.from(cipherText.substring(cipherText.length - 32), 'hex')
        // remove the auth tag from the cipher text
        cipherText = cipherText.substring(0, cipherText.length - 32)
    }

    // create decipheriv using the secret key and the initialization vector
    const cipher = createDecipheriv(mode, key, iv)
    if (authMode) {
        cipher.setAuthTag(authTag)
    }

    // update decipher with the cipher text
    let clearText = cipher.update(cipherText, 'hex', 'utf8');
    // finalize decipher
    clearText += cipher.final()

    return clearText
}
